[triggers]: Added tutorial describing how to register a trigger. (#35)
hyperledger/iroha-2-docs: src/guide/advanced/triggers.md

where the `TriggerId` is a simple wrapper around a single `Name`, i.e. string with no white-space and no reserved characters (`@`, `#`). In the future, we shall add scoped triggers, and the Id will be expanded to be either a global trigger, or a trigger with a domain name. This is what determines the scope of the trigger.

> `Metadata` is the same kind of `Metadata` that can be attached to accounts, domains, assets and even transactions. In the future, such `Metadata` would also be possible to use as a persistent key-value store unique to the trigger, alongside `KeyValue` assets.

Reviewer: ```suggestion
`Metadata` is the same kind of `Metadata` that can be attached to accounts, domains, assets, and even transactions. In the future, such `Metadata` would also have the possibility to be used as a persistent key-value store unique to the trigger, alongside `KeyValue` assets.
```

> Each of these shall be covered in its own right, and some information shall be provided on how to use each of them in detail. Both this tutorial and triggers themselves are under construction; some triggers don't exist yet, the API of others will change drastically in the following release. We shall do our best to describe all of what we can, with as much detail as we can, and clearly signpost which parts of this tutorial will be made obsolete in the next release.

Reviewer: ```suggestion
Each of these shall be covered in its own right, and some information shall be provided on how to use each of them in detail. Both this tutorial and triggers themselves are under construction; some triggers don't exist yet, while the API of others will change drastically in the following release. We shall do our best to describe all of what we can, with as much detail as we can, and clearly signpost which parts of this tutorial will be made obsolete in the next release.
```

So, the offset is meant to anticipate when the block would get added to the chain, so that people who were just 4 minutes early don't have to be potentially several hours late. Because more triggers get executed sooner, your throughput is also infinitesimally smaller. We could also say "you should aim to execute your trigger slightly earlier than consensus starts", but people writing smartcontracts already have too much to worry about.

### Pre-commit Triggers

> This is a variant of timed triggers that gets run before transactions get committed. In reality it triggers in the next block, by leaving a special event, but it is effectively a delayed pre-commit that can track behaviour of transactions in the pipeline.

Reviewer: ```suggestion
This is a variant of timed triggers that gets run before transactions get committed. In reality, it triggers in the next block, by leaving a special event, but it is effectively a delayed pre-commit that can track the behaviour of transactions in the pipeline.
```

> This is a variant of timed triggers that gets run before transactions get committed. In reality it triggers in the next block, by leaving a special event, but it is effectively a delayed pre-commit that can track the behaviour of transactions in the pipeline.

Reviewer: ```suggestion
This is a variant of timed triggers that gets run before transactions get committed. In reality, it triggers in the next block by leaving a special event, but it is effectively a delayed pre-commit that can track the behavior of transactions in the pipeline.
```

::: info

These triggers are not meant for restricting the execution of transactions. If you want to stop your users from transferring more than X amount of Y to user Z, you really want a _permission_. While you could hack the pre-commit triggers to emulate the desired behaviour, this is neither economical in terms of gas fees, nor computation.

Until Iroha 2 supports WASM-based _permissions validators_, however, your only choice are pre-commit triggers.

:::

### Block-based Triggers

Same as timed triggers, but instead of only being interested in the time-stamp, this trigger is interested in the block height. While the mechanism for the triggers is similar the use cases are different.

### By-call Triggers

> These triggers get executed only once the `CallTrigger(trigger_name)` instruction is executed. These can be useful if you want to achieve dynamic linkage between different smartcontract modules. Space is precious, so want to use as little of it as you can. So you follow the UNIX design philosophy, and instead of creating one large smartcontract, you create many smaller ones, and re-use as much logic as you can.

Reviewer: ```suggestion
These triggers get executed only once the `CallTrigger(trigger_name)` instruction is executed. These can be useful if you want to achieve dynamic linkage between different smartcontract modules. Space is precious, so you want to use as little of it as you can. Therefore, you follow the UNIX design philosophy, and instead of creating one large smartcontract, you create many smaller ones, and re-use as much logic as you can.
```

### Why not WASM

> The above observation can be generalised. WASM can do any logic that a Turing complete machine could using the data available via queries. So in theory for event-based triggers, you could create an `AcceptAll` event filter and do all of the processing using the key-value store as persistent storage, and determining if you want to execute using easy to understand Rust code, and not our admittedly cumbersome, `EventFilters`. We don't want that. WASM takes up significantly more space, and takes longer to execute compared to plain ISI, which are slower than `EventFilters`. We want you to want to use the `EventFilters` because they would make the process much more efficient, and are working tirelessly to make the experience of using event filters much less cumbersome.

Reviewer: ```suggestion
The above observation can be generalised. WASM can do any logic that a Turing complete machine could, using the data available via queries. So in theory for event-based triggers, you could create an `AcceptAll` event filter and do all of the processing using the key-value store as persistent storage, then determining if you want to execute using easy-to-understand Rust code, and not our admittedly cumbersome, `EventFilters`. We don't want that. WASM takes up significantly more space and takes longer to execute compared to plain ISI, which are slower than `EventFilters`. We want you to want to use the `EventFilters` because they would make the process much more efficient, and we are working tirelessly to make the experience of using event filters much less cumbersome.
```

> However, as was mentioned on several occasions previously, implementing a feature well takes time and effort. Ergonomics must be balanced against safety and reliability, so we cannot just make things more easy to use. We want them to retain many of the advantages of strong typing.

Reviewer: ```suggestion
However, as was mentioned previously on several occasions, implementing a feature properly takes time and effort. Ergonomics must be balanced against safety and reliability, so we cannot just make things easier to use. We want them to retain many of the advantages of strong typing.
```
